gen/field.go

```diff
@@ -94,23 +105,88 @@ func (f fieldGroupGenerator) DefineStruct(g Generator) error {
 			<end>
 		}`,
 		f,
-		TemplateFunc("tag", func(f *compile.FieldSpec) string {
-			// We want to add omitempty if the field is an optional struct or
-			// primitive to redure "null" noise. We won't add omitempty for
-			// optional collections because omitempty doesn't differentiate
-			// between nil and empty collections.
-
-			if (isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.Required {
-				return fmt.Sprintf("`json:\"%s,omitempty\"`", f.Name)
-			}
-
-			return fmt.Sprintf("`json:%q`", f.Name)
-			// TODO(abg): Take go.tag and js.name annotations into account
-		}),
+		TemplateFunc("tag", generateTags),
 		TemplateFunc("declFieldName", f.declFieldName),
 	)
 }
 
+// generateTags parses the annotation on the thrift field and creates the resulting go tag
+func generateTags(f *compile.FieldSpec) string {
+	// start with default json tag
+	tk := []string{jsonTagName} // list of tag keys to preserve order
+	ts := map[string]string{
+		jsonTagName: compileJSONTagValue(f, f.Name),
+	}
+
+	// process go tags and overwrite json tag if specified in thrift annotation
+	if goAnnotation := f.Annotations[goTagName]; goAnnotation != "" {
+		goTags, goTagKeys := mapFromRawTag(goAnnotation)
+		for _, k := range goTagKeys {
+			if _, exists := ts[k]; !exists {
+				tk = append(tk, k)
+			}
+			ts[k] = goTags[k]
+		}
+
+		if jt := reflect.StructTag(goAnnotation).Get(jsonTagName); jt != "" {
+			ts[jsonTagName] = compileJSONTagValue(f, jt)
+		}
+	}
+	// TODO (abg): Take js.name annotation into account
+	return mapToRawTag(ts, tk)
+}
+
+// mapFromRawTag constructs a tag's key, value map from the raw tag string
+// `json:"foo,required" bson:"foo bar"` -> {"json": "foo,required", "bson": "foo bar"}
+// it also returns a list of keys to preserve tag order
+func mapFromRawTag(a string) (map[string]string, []string) {
+	ts, tk := map[string]string{}, []string{}
+	split := strings.Split(a, " ")
+	for i, t := range split {
+		if t[len(t)-1] != '"' {
+			split[i+1] = strings.Join(split[i:i+2], " ")
+			continue
+		}
+		k := strings.Split(t, ":")[0]
+		if _, exists := ts[k]; !exists {
+			tk = append(tk, k)
+		}
+		ts[k] = reflect.StructTag(t).Get(k)
+	}
+	return ts, tk
+}
+
+// mapToRawTag constructs a raw tag string from a key, value map
+// {"json": "foo,required", "bson": "bar"} -> `json:"foo,required" bson:"bar"`
+// it also consumes a list of keys to preserve tag order
+func mapToRawTag(t map[string]string, tk []string) string {
+	var a []string
+	for _, k := range tk {
+		a = append(a, fmt.Sprintf("%s:%q", k, t[k]))
+	}
+	return fmt.Sprintf("`%s`", strings.Join(a, " "))
+}
+
+// compileJSONTagValue constructs the json tag value
+// it inspects the thrift field spec and sets the tag value accordingly
+func compileJSONTagValue(f *compile.FieldSpec, segments ...string) string {
```

Comment:
nit: you're not really using the varargs here. Just accept `name` as a single argument. Plus you lose some safety because the code below assumes `len(segments) > 0`.